netebpfext/net_ebpf_ext.c

```diff
@@ -613,123 +607,31 @@ _net_ebpf_ext_no_op_flow_delete(uint16_t layer_id, uint32_t fwpm_callout_id, uin
-    _IRQL_requires_(DISPATCH_LEVEL) _IRQL_requires_same_ VOID _net_ebpf_ext_rundown(
-        _In_ KDPC* dpc,
-        _In_opt_ void* deferred_context,
-        _In_opt_ void* system_argument_1,
-        _In_opt_ void* system_argument_2)
-{
-    net_ebpf_ext_hook_provider_registration_t* registration =
-        (net_ebpf_ext_hook_provider_registration_t*)deferred_context;
-
-    UNREFERENCED_PARAMETER(dpc);
-    UNREFERENCED_PARAMETER(system_argument_1);
-    UNREFERENCED_PARAMETER(system_argument_2);
-    if (registration)
-        KeSetEvent(&registration->rundown_wait, 0, FALSE);
-}
-
-static void
-_net_ebpf_ext_init_rundown(_In_ net_ebpf_ext_hook_provider_registration_t* registration)
-{
-    KeInitializeEvent(&(registration->rundown_wait), NotificationEvent, FALSE);
-    KeInitializeDpc(&(registration->rundown_dpc), _net_ebpf_ext_rundown, registration);
-}
-
-static void
-_net_ebpf_ext_wait_for_rundown(_In_ net_ebpf_ext_hook_provider_registration_t* registration)
-{
-    // Queue a DPC to each CPU and wait for it to run.
-    // After it has run on each CPU we can be sure that no
-    // DPC is busy processing a hook.
-    uint32_t maximum_processor = KeQueryMaximumProcessorCount();
-    uint32_t processor;
-    for (processor = 0; processor < maximum_processor; processor++) {
-        KeSetTargetProcessorDpc(&registration->rundown_dpc, (uint8_t)processor);
-        if (KeInsertQueueDpc(&registration->rundown_dpc, NULL, NULL)) {
-            KeWaitForSingleObject(&registration->rundown_wait, Executive, KernelMode, FALSE, NULL);
-        }
-    }
-}
-
-ebpf_result_t
-_net_ebpf_ext_provider_client_detach_callback(void* context, const GUID* client_id)
-{
-    net_ebpf_ext_hook_provider_registration_t* hook_registration = (net_ebpf_ext_hook_provider_registration_t*)context;
-    UNREFERENCED_PARAMETER(client_id);
-    hook_registration->client_binding_context = NULL;
-    hook_registration->client_data = NULL;
-    _net_ebpf_ext_wait_for_rundown(hook_registration);
-    hook_registration->invoke_hook = NULL;
-    return EBPF_SUCCESS;
-}
-
 NTSTATUS
 net_ebpf_ext_register_providers()
 {
     ebpf_result_t return_value;
-    _net_ebpf_ext_init_rundown(&_ebpf_xdp_hook_provider_registration);
-    _net_ebpf_ext_init_rundown(&_ebpf_bind_hook_provider_registration);
-
-    return_value = ebpf_provider_load(
-        &_ebpf_xdp_hook_provider_registration.provider,
-        &EBPF_ATTACH_TYPE_XDP,
-        &_ebpf_xdp_hook_provider_registration,
-        _ebpf_xdp_hook_provider_registration.provider_data,
-        NULL,
-        &_ebpf_xdp_hook_provider_registration,
-        _net_ebpf_ext_provider_client_attach_callback,
-        _net_ebpf_ext_provider_client_detach_callback);
-
+    return_value = ebpf_ext_attach_register_provider(
+        &EBPF_ATTACH_TYPE_XDP, EBPF_EXT_HOOK_EXECUTION_DISPATCH, &_ebpf_xdp_hook_provider_registration);
     if (return_value != EBPF_SUCCESS) {
-        goto Done;
+        return STATUS_UNSUCCESSFUL;
```

Comment:
ebpfcore.sys has _ebpf_result_to_ntstatus(), should we export that for use by extensions like here?

Reply:
Issue #267 filed